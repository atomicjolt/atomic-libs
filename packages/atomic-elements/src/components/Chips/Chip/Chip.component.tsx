import React, { useContext } from "react";
import { filterDOMProps, mergeProps, useObjectRef } from "@react-aria/utils";
import { useTag } from "@react-aria/tag";
import { createLeafComponent } from "@react-aria/collections";

import { IconButton } from "@components/Buttons/IconButton";
import { useConditionalPress } from "@hooks/useConditionalPress";
import { useFocusRing } from "@hooks/useFocusRing";
import { useContextProps } from "@hooks/useContextProps";
import { useRenderProps } from "@hooks";
import { ChipGroupStateContext } from "@components/Fields/ChipGroupField/ChipGroupField.context";

import { ChipArgs, ChipGroupChipProps, ChipInternalProps } from "./Chip.types";
import { ChipContent, ChipWrapper } from "./Chip.styles";
import { ChipContext } from "./Chip.context";

export function ChipLeaf<T>(...args: ChipArgs<T>) {
  const [props, ref] = useContextProps(ChipContext, args[0], args[1]);

  // We're being rendered standalone
  if (args.length === 2) {
    return (
      <ChipInternal {...props} ref={ref} allowsRemoving={!!props.onRemove} />
    );
  }

  // We're being rendered as part of a collection (i.e ChipGroup)
  const item = args[2];
  return <ChipGroupChip item={item} itemRef={ref} {...props} />;
}

/**
 * Chip component. Can be used stand-alone, or within a parent
 * `ChipGroup`
 */
export const Chip = createLeafComponent("item", ChipLeaf);

function ChipGroupChip<T>(props: ChipGroupChipProps<T>) {
  const { item } = props;
  const ref = useObjectRef<HTMLDivElement>(props.itemRef);
  const state = useContext(ChipGroupStateContext)!;
  const {
    rowProps,
    gridCellProps,
    removeButtonProps,
    allowsRemoving,
    isFocused,
    isSelected,
  } = useTag(props, state, ref);

  const { focusProps, isFocusVisible } = useFocusRing({ within: true });

  const isDisabled = state.disabledKeys.has(item.key);

  const renderProps = useRenderProps({
    componentClassName: "aje-chip",
    values: {
      isSelected,
      isFocusVisible,
      isFocused,
    },
    ...props,
  });

  return (
    <ChipWrapper
      ref={ref}
      {...mergeProps(rowProps, focusProps)}
      {...renderProps}
    >
      <ChipContent {...gridCellProps}>
        {renderProps.children}
        {allowsRemoving && (
          <IconButton
            icon="close"
            size="small"
            variant="chip"
            {...mergeProps(removeButtonProps, { isDisabled })}
          />
        )}
      </ChipContent>
    </ChipWrapper>
  );
}

export const ChipInternal = React.forwardRef(function ChipInternal<T>(
  props: ChipInternalProps<T>,
  ref: React.Ref<HTMLDivElement>
) {
  const {
    className,
    variant,
    onRemove,
    isDisabled,
    children,
    allowsRemoving = false,
    ...rest
  } = props;

  const { pressProps } = useConditionalPress(rest);

  const renderProps = useRenderProps({
    componentClassName: "aje-chip",
    values: {
      isSelected: false,
      isFocusVisible: false,
      isFocused: false,
    },
    ...props,
  });

  return (
    <ChipWrapper
      ref={ref}
      {...mergeProps({ "aria-disabled": isDisabled || undefined }, pressProps)}
      {...renderProps}
      {...filterDOMProps(props as any)}
    >
      <ChipContent>
        {renderProps.children}
        {allowsRemoving && (
          <IconButton
            icon="close"
            size="small"
            variant="chip"
            isDisabled={isDisabled}
            onPress={onRemove}
          />
        )}
      </ChipContent>
    </ChipWrapper>
  );
});
