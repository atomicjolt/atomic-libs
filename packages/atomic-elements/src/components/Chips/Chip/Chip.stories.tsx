import { Meta, StoryObj } from "@storybook/react";
import { Chip } from ".";
import { fn } from "@storybook/test";
import { PressableArgTypes } from "@sb/helpers";

export default {
  title: "Chips/Chip",
  component: Chip,
  parameters: {
    layout: "centered",
  },
  argTypes: {
    children: {
      type: "string",
    },
    variant: {
      control: "select",
      options: ["default", "warning", "success", "danger"],
    },
    onRemove: {
      description: "Function to call when the chip is removed",
      table: {
        category: "Events",
      },
    },
  },
} as Meta<typeof Chip>;

type Story = StoryObj<typeof Chip>;

export const Primary: Story = {
  args: {
    children: "Chip",
  },
};

export const Removable: Story = {
  args: {
    ...Primary.args,
    onRemove: fn(),
  },
};

export const Pressasble: Story = {
  args: {
    ...Primary.args,
    onPress: fn(),
  },
  argTypes: {
    ...PressableArgTypes,
  },
};
