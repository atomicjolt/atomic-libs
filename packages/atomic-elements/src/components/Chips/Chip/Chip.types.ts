import { Node } from "react-stately";
import { AriaLabelProps, DomProps, SuggestStrings } from "../../../types";
import { PressEvent, PressProps } from "@react-aria/interactions";
import { ItemProps } from "../../Collection";

type ChipVariants = SuggestStrings<
  "default" | "warning" | "success" | "danger"
>;

export interface ChipProps<T> extends ItemProps, PressProps {
  variant?: ChipVariants;
  /** Handler that is called when the user
   * clicks the remove button for the chip */
  onRemove?: (e: PressEvent) => void;
  isDisabled?: boolean;
}

export type ChipArgs<T> =
  | [ChipProps<T>, React.ForwardedRef<HTMLDivElement>]
  | [ChipProps<T>, React.ForwardedRef<HTMLDivElement>, Node<T>];

export interface ChipGroupChipProps<T> extends ChipProps<T> {
  item: Node<T>;
  itemRef: React.ForwardedRef<HTMLDivElement>;
}

export interface ChipInternalProps<T> extends ChipProps<T> {
  allowsRemoving?: boolean;
}
