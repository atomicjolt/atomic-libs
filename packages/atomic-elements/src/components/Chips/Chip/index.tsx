export { Chip, ChipLeaf } from "./Chip.component";
export type { ChipProps } from "./Chip.types";
