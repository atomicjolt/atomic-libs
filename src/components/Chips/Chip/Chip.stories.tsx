import { Meta, StoryObj } from "@storybook/react";
import { Chip } from ".";
import { fn } from "@storybook/test";

export default {
  title: "Chips/Chip",
  component: Chip,
  parameters: {
    layout: "centered",
  },
  argTypes: {
    onRemove: {
      description: "Function to call when the chip is removed",
    },
    children: {
      type: "string",
    },
  },
} as Meta<typeof Chip>;

type Story = StoryObj<typeof Chip>;

export const Primary: Story = {
  args: {
    children: "Chip",
  },
};

export const Removable: Story = {
  args: {
    ...Primary.args,
    onRemove: fn(),
  },
};

export const Pressasble: Story = {
  args: {
    ...Primary.args,
    onPress: fn(),
  },
};
