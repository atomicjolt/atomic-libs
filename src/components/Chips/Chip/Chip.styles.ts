import styled from "styled-components";
import mixins from "@/styles/mixins";

export const ChipContent = styled.div`
  ${mixins.Regular}
  min-height: 2.2rem;
  padding: 0.3rem 0.6rem;
  line-height: 1;
  font-size: 1.2rem;
  font-weight: var(--font-weight-bold);
  display: flex;
  align-items: center;
  gap: 0.6rem;

  button {
    height: 1.6rem;
    width: 1.6rem;
    padding: 0;
    background: none;
    border: none;
    position: relative;
    isolation: isolate;
    margin-right: -3px;

    i {
      color: var(--chip-text-clr);
    }

    &::before {
      content: "";
      position: absolute;
      inset: -3px;
      border-radius: var(--chip-radius);
      background-color: var(--chip-bg-color);
      z-index: -1;
    }

    &:hover {
      cursor: pointer;
      &::before {
        background-color: var(--chip-btn-hover-bg-clr);
      }
    }
  }
`;

export const ChipWrapper = styled.div`
  display: inline-block;
  background-color: var(--chip-bg-clr);
  color: var(--chip-text-clr);
  border-radius: var(--chip-radius);

  &.aje-chip--warning {
    --chip-bg-clr: var(--warning100);
    --chip-hover-bg-clr: var(--warning200);
    --chip-btn-hover-bg-clr: var(--warning200);
  }

  &.aje-chip--success {
    --chip-bg-clr: var(--success100);
    --chip-hover-bg-clr: var(--success200);
    --chip-btn-hover-bg-clr: var(--success200);
  }

  &[role="button"]:hover,
  &[aria-selected]:hover {
    cursor: pointer;
    --chip-bg-clr: var(--chip-hover-bg-clr);
  }

  /* For some reason, when you click the remove button on a
    chip, the focus moves to the last keyboard-focused element.
    BUT it doesn't set the focus-visible attribute to true.
    So we have to setup the outline twice
  */
  &:focus-visible {
    outline: none;
  }

  &[data-focus-visible="true"] {
    outline: var(--outline);

    &:focus-within {
      outline: var(--outline);
    }
  }

  &[aria-selected="true"] {
    --chip-bg-clr: var(--chip-selected-bg-clr);
    --chip-text-clr: var(--chip-selected-text-clr);

    &:hover {
      --chip-hover-bg-clr: var(--chip-selected-hover-bg-clr);
    }
  }

  &[aria-disabled] {
    opacity: 0.5;

    ${ChipContent} button::hover {
      &::before {
        background-color: var(--neutral100);
      }
    }
  }
`;
