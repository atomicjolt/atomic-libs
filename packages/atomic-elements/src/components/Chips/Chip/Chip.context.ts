import { createComponentContext } from "@utils/index";
import { ChipProps } from "./Chip.types";

export const ChipContext = createComponentContext<ChipProps<any>>();
