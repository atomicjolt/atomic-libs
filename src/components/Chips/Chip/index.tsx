import React from "react";
import classNames from "classnames";
import type { ItemProps } from "react-stately";
import {
  AriaButtonProps,
  PressEvent,
  PressProps,
  mergeProps,
} from "react-aria";
import { copyStaticProperties } from "@/clone";
import { useVariantClass } from "@/hooks";
import { HasClassName, SuggestStrings } from "@/types";
import { Item } from "@/components/Collection";
import IconButton from "@/components/Buttons/IconButton";
import { useConditionalPress } from "@/hooks/useConditionalPress";
import { ChipContent, ChipWrapper } from "./Chip.styles";

type ChipVariants = SuggestStrings<"default" | "warning" | "success">;

export interface ChipProps<T> extends ItemProps<T>, PressProps, HasClassName {
  children: React.ReactNode;
  variant?: ChipVariants;
  /** Handler that is called when the user
   * clicks the remove button for the chip */
  onRemove?: (e: PressEvent) => void;
  isDisabled?: boolean;
}

/** Chip component */
export function Chip<T>(props: ChipProps<T>) {
  return <ChipInternal {...props} allowsRemoving={!!props.onRemove} />;
}

copyStaticProperties(Item, Chip);

interface ChipInternalProps<T> extends ChipProps<T> {
  wrapperProps?: React.DOMAttributes<HTMLDivElement>;
  contentProps?: React.DOMAttributes<HTMLDivElement>;
  removeButtonProps?: AriaButtonProps<"button">;
  allowsRemoving?: boolean;
}

export const ChipInternal = React.forwardRef<
  HTMLDivElement,
  ChipInternalProps<any>
>(function ChipInternal<T>(
  props: ChipInternalProps<T>,
  ref: React.Ref<HTMLDivElement>
) {
  const {
    className,
    variant = "default",
    onRemove,
    isDisabled,
    children,
    wrapperProps = {},
    contentProps = {},
    removeButtonProps = {},
    allowsRemoving = false,
    ...rest
  } = props;

  const variantClass = useVariantClass("aje-chip", variant);
  const { pressProps } = useConditionalPress(rest);

  const allWrapperProps = [
    wrapperProps,
    { "aria-disabled": isDisabled || undefined },
  ];

  if (!isDisabled) {
    allWrapperProps.push(pressProps);
  }

  return (
    <ChipWrapper
      className={classNames("aje-chip", variantClass, className)}
      ref={ref}
      {...mergeProps(...allWrapperProps)}
    >
      <ChipContent {...contentProps}>
        {children}

        {allowsRemoving && (
          <IconButton
            icon="close"
            size="small"
            variant="content"
            {...mergeProps(
              {
                isDisabled,
                onPress: onRemove,
              },
              removeButtonProps
            )}
          />
        )}
      </ChipContent>
    </ChipWrapper>
  );
});
