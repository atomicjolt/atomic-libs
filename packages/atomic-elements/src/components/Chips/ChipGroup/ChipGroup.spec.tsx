import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { ChipGroup } from ".";
import { Chip } from "../Chip";

describe("ChipGroup", () => {
  test("matches snapshot", () => {
    const result = render(
      <ChipGroup label="My Chip Group">
        <Chip>Item</Chip>
      </ChipGroup>
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("renders label", () => {
    render(
      <ChipGroup label="My Chip Group">
        <Chip>Item</Chip>
      </ChipGroup>
    );

    expect(screen.getByLabelText("My Chip Group")).not.toBeNull();
  });

  test("renders chips", () => {
    const chips = [
      { id: "1", rendered: "Chip 1" },
      { id: "2", rendered: "Chip 2" },
      { id: "3", rendered: "Chip 3" },
    ];
    render(
      <ChipGroup items={chips} label="Label">
        {({ rendered }) => <Chip>{rendered}</Chip>}
      </ChipGroup>
    );
    chips.forEach((chip) => {
      expect(screen.getByText(chip.rendered)).not.toBeNull();
    });
  });
});
