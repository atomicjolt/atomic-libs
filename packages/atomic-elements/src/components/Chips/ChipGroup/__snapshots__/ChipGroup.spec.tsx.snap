// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChipGroup > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-ovuCP eiAkiR aje-chip-group"
  >
    <label
      class="sc-blHHSb fcIAmr"
      id="react-aria-:r1:"
    >
      My Chip Group
    </label>
    <div
      aria-atomic="false"
      aria-labelledby="react-aria-:r1:"
      aria-live="off"
      aria-relevant="additions"
      class="sc-eOzmre foFCjW"
      id="react-aria-:r0:"
      role="grid"
      tabindex="0"
    >
      <div
        aria-label="Item"
        class="sc-fFoeYl kvHja-d aje-chip"
        data-key="react-aria-1"
        id="react-aria-:r0:-react-aria-1"
        role="row"
        tabindex="0"
      >
        <div
          aria-colindex="1"
          class="sc-eauhAA hNmxTV"
          role="gridcell"
        >
          Item
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
